File: app/src/main/java/com/dleibovych/epictale/fragment/DiaryFragment.java
package com.dleibovych.epictale.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.dleibovych.epictale.DataViewMode;
import com.dleibovych.epictale.R;
import com.dleibovych.epictale.TheTaleApplication;
import com.dleibovych.epictale.api.ApiResponseCallback;
import com.dleibovych.epictale.api.model.DiaryEntry;
import com.dleibovych.epictale.api.request.GameInfoRequest;
import com.dleibovych.epictale.api.response.GameInfoResponse;
import com.dleibovych.epictale.util.PreferencesManager;
import com.dleibovych.epictale.util.RequestUtils;
import com.dleibovych.epictale.util.UiUtils;

import java.net.CookieManager;

import javax.inject.Inject;

import okhttp3.OkHttpClient;

/**
 * @author Hamster
 * @since 06.10.2014
 */
public class DiaryFragment extends WrapperFragment {

    @Inject OkHttpClient client;
    @Inject CookieManager manager;

    private LayoutInflater layoutInflater;

    private View rootView;

    private ViewGroup diaryContainer;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        TheTaleApplication
                .getComponentProvider()
                .getAppComponent()
                .inject(this);


        layoutInflater = inflater;
        rootView = inflater.inflate(R.layout.fragment_diary, container, false);

        diaryContainer = rootView.findViewById(R.id.diary_container);

        return wrapView(layoutInflater, rootView);
    }

    @Override
    public void refresh(final boolean isGlobal) {
        super.refresh(isGlobal);

        final ApiResponseCallback<GameInfoResponse> callback = RequestUtils.wrapCallback(new ApiResponseCallback<GameInfoResponse>() {
            @Override
            public void processResponse(GameInfoResponse response) {
                diaryContainer.removeAllViews();
                for(int i = response.account.hero.diary.size() - 1; i >= 0; i--) {
                    final DiaryEntry diaryEntry = response.account.hero.diary.get(i);
                    final View diaryEntryView = layoutInflater.inflate(R.layout.item_diary, diaryContainer, false);
                    UiUtils.setText(
                            diaryEntryView.findViewById(R.id.diary_place),
                            diaryEntry.place);
                    UiUtils.setText(
                            diaryEntryView.findViewById(R.id.diary_time),
                            String.format("%s %s", diaryEntry.time, diaryEntry.date));
                    UiUtils.setText(
                            diaryEntryView.findViewById(R.id.diary_text),
                            diaryEntry.text);
                    diaryContainer.addView(diaryEntryView);
                }
                setMode(DataViewMode.DATA);
            }

            @Override
            public void processError(GameInfoResponse response) {
                setError(response.errorMessage);
            }
        }, this);

        final int watchingAccountId = PreferencesManager.getWatchingAccountId();
        if(watchingAccountId == 0) {
            new GameInfoRequest(client, manager, true).execute(callback, true);
        } else {
            new GameInfoRequest(client, manager, true).execute(watchingAccountId, callback, true);
        }
    }

}

File: app/src/main/java/com/dleibovych/epictale/game/diary/DiaryFragment.kt
package com.dleibovych.epictale.game.diary

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.dleibovych.epictale.R
import com.dleibovych.epictale.game.di.GameComponentProvider
import com.dleibovych.epictale.util.UiUtils
import kotlinx.android.synthetic.main.fragment_diary.*

import javax.inject.Inject

import org.thetale.api.models.DiaryInfo

class DiaryFragment : Fragment(), DiaryView {

    @Inject
    lateinit var presenter: DiaryPresenter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        (activity!!.application as GameComponentProvider)
                .provideGameComponent()
                ?.inject(this)

        presenter.view = this

        return inflater.inflate(R.layout.fragment_diary, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        error.onRetryClick(View.OnClickListener {
            presenter.loadDiary()
        })
    }

    override fun onDestroyView() {
        super.onDestroyView()

        presenter.view = null
    }

    override fun onStart() {
        super.onStart()

        presenter.start()
    }

    override fun onStop() {
        super.onStop()

        presenter.stop()
    }

    override fun onDestroy() {
        super.onDestroy()

        if (activity!!.isFinishing) {
            presenter.dispose()
        }
    }

    override fun showProgress() {
        progress.visibility = View.VISIBLE
        diaryContainer.visibility = View.GONE
        error.visibility = View.GONE
    }

    override fun showDiary(info: DiaryInfo) {
        diaryContainer.visibility = View.VISIBLE
        progress.visibility = View.GONE
        error.visibility = View.GONE

        diaryContainer!!.removeAllViews()
        for (message in info.messages) {
            val diaryEntryView = layoutInflater!!.inflate(R.layout.item_diary, diaryContainer, false)
            UiUtils.setText(
                    diaryEntryView.findViewById(R.id.diary_place),
                    message.position)
            UiUtils.setText(
                    diaryEntryView.findViewById(R.id.diary_time),
                    String.format("%s %s", message.gameTime, message.gameDate))
            UiUtils.setText(
                    diaryEntryView.findViewById(R.id.diary_text),
                    message.message)
            diaryContainer!!.addView(diaryEntryView)
        }
    }

    override fun showError() {
        diaryContainer.visibility = View.GONE
        progress.visibility = View.GONE
        error.visibility = View.VISIBLE

        error.setErrorText(getString(R.string.common_error))
    }
}
